pkg/providers/vsphere/session/session_vm_update.go

```diff
@@ -56,6 +62,355 @@ type VMResizeArgs struct {
+		conditions.Delete(vmCtx.VM, vmopv1.GuestIDReconfiguredCondition)
+	}
+
+	networkResults, err := s.reconcileNetworkInterfaces(vmCtx)
+	if err != nil {
+		return err
+	}
+
+	switch vmCtx.MoVM.Runtime.PowerState {
+	case vimtypes.VirtualMachinePowerStatePoweredOn:
+
+		if err := s.poweredOnReconfigure(
+			vmCtx,
+			vcVM,
+			vmCtx.MoVM.Config); err != nil {
+
+			return err
+		}
+
+	case vimtypes.VirtualMachinePowerStatePoweredOff:
+
+		if f := pkgcfg.FromContext(vmCtx).Features; f.VMResize ||
+			f.VMResizeCPUMemory {
+
+			resizeArgs.NetworkResults = networkResults
+			if err := s.resizeVMWhenPoweredStateOff(
+				vmCtx,
+				vcVM,
+				vmCtx.MoVM,
+				resizeArgs); err != nil {
+
+				return err
+			}
+
+		} else {
+
+			updateArgs.NetworkResults = networkResults
+			if err := s.poweredOffReconfigure(
+				vmCtx,
+				vcVM,
+				vmCtx.MoVM.Config,
+				updateArgs); err != nil {
+
+				return err
+			}
+		}
+	}
+
+	return s.reconcileNetworkAndGuestCustomizationState(
+		vmCtx,
+		vcVM,
+		bootstrapData,
+		networkResults)
+}
+
+func (s *Session) poweredOnReconfigure(
+	vmCtx pkgctx.VirtualMachineContext,
+	vcVM *object.VirtualMachine,
+	config *vimtypes.VirtualMachineConfigInfo) error {
+
+	configSpec := &vimtypes.VirtualMachineConfigSpec{}
+
+	if err := vmopv1util.OverwriteAlwaysResizeConfigSpec(
+		vmCtx,
+		*vmCtx.VM,
+		*config,
+		configSpec); err != nil {
+
+		return err
+	}
+
+	UpdateConfigSpecExtraConfig(vmCtx, config, configSpec, vmCtx.VM, nil)
+	UpdateConfigSpecChangeBlockTracking(vmCtx, config, configSpec, vmCtx.VM.Spec)
+
+	if err := virtualmachine.UpdateConfigSpecCdromDeviceConnection(vmCtx, s.Client.RestClient(), s.K8sClient, config, configSpec); err != nil {
+		return fmt.Errorf("update CD-ROM device connection error: %w", err)
+	}
+
+	if err := doReconfigure(
+		logr.NewContext(
+			vmCtx,
+			vmCtx.Logger.WithName("poweredOnReconfigure"),
+		),
+		s.K8sClient,
+		vmCtx.VM,
+		vcVM,
+		vmCtx.MoVM,
+		*configSpec); err != nil {
+
+		return err
+	}
+
+	if err := s.reconcileChangeTracking(vmCtx, configSpec); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Session) poweredOffReconfigure(
+	vmCtx pkgctx.VirtualMachineContext,
+	vcVM *object.VirtualMachine,
+	config *vimtypes.VirtualMachineConfigInfo,
+	updateArgs *VMUpdateArgs) error {
+
+	var configSpec *vimtypes.VirtualMachineConfigSpec
+	var needsResize bool
+	var err error
+
+	if pkgcfg.FromContext(vmCtx).Features.VMResize {
+		configSpec, needsResize, err = s.getResizeConfigSpecForPoweredOffVM(
+			vmCtx, config, updateArgs)
+	} else {
+		configSpec, needsResize, err = s.getConfigSpecForPoweredOffVM(
+			vmCtx, config, updateArgs)
+	}
+	if err != nil {
+		return err
+	}
+
+	reconfigErr := doReconfigure(
+		logr.NewContext(
+			vmCtx,
+			vmCtx.Logger.WithName("poweredOffReconfigure"),
+		),
+		s.K8sClient,
+		vmCtx.VM,
+		vcVM,
+		vmCtx.MoVM,
+		*configSpec)
+
+	if needsResize && errors.Is(reconfigErr, ErrReconfigure) {
```

Comment:
I think the later check should be `(reconfigErr == nil || errors.Is(reconfigErr, ErrReconfigure))` because it is still a resize even if the new class doesn't result in a reconfigure.